File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Go/f5eedfa3d3ba232ace3697125d6ca9fc.go
package main

import (
	"fmt"
)

const fftmod = 1000000007

func cnk(n, k int) int64 {
	fact := make([]int64, n+1)
	revFact := make([]int64, n+1)
	fact[0] = 1
	for i := 1; i <= n; i++ {
		fact[i] = fact[i-1] * int64(i) % fftmod
	}
	revFact[n] = powmod(fact[n], fftmod-2, fftmod)
	for i := n - 1; i >= 0; i-- {
		revFact[i] = revFact[i+1] * int64(i+1) % fftmod
	}
	return fact[n] * revFact[k] % fftmod * revFact[n-k] % fftmod
}

func powmod(x, n, m int64) int64 {
	r := int64(1)
	x %= m
	for n > 0 {
		if n&1 == 1 {
			r = r * x % m
		}
		x = x * x % m
		n >>= 1
	}
	return r
}

func solve() int {
	var n, k int
	fmt.Scan(&n, &k)
	a := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&a[i])
	}
	a = append(a, a[0])
	cnt := 0
	for i := 0; i < n; i++ {
		if a[i] != a[i+1] {
			cnt++
		}
	}
	if k == 1 || n == 1 {
		fmt.Println(0)
		return 0
	}
	ans := int64(0)
	freeChoices := k - 2
	for i := 0; i <= cnt/2; i++ {
		cur := cnk(cnt, i) * cnk(cnt-i, i) % fftmod * powmod(freeChoices, cnt-2*i, fftmod) % fftmod * powmod(int64(k), int64(n-cnt), fftmod) % fftmod
		ans += cur
	}
	ans %= fftmod
	tot := powmod(int64(k), int64(n), fftmod) - ans
	if tot < 0 {
		tot += fftmod
	}
	fmt.Println(tot * powmod(2, fftmod-2, fftmod) % fftmod)
	return 0
}

func main() {
	solve()
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/C/f5eedfa3d3ba232ace3697125d6ca9fc.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <math.h>

#define MAX_N 1010101
#define MAX_K 202020
#define MOD 1000000007

int a[MAX_N];
int fact[MAX_K];
int revFact[MAX_K];

long long cnk(int n, int k) {
    return (fact[n] * 1LL * revFact[k] % MOD * revFact[n - k] % MOD);
}

int solve() {
    fact[0] = 1;
    const int N = 200000;
    for (int i = 1; i <= N; ++i) {
        fact[i] = (fact[i - 1] * 1LL * i) % MOD;
    }
    revFact[N] = powmod(fact[N], MOD - 2, MOD);
    for (int i = N - 1; i >= 0; --i) {
        revFact[i] = (revFact[i + 1] * (i + 1LL)) % MOD;
    }
    
    int n, k;
    scanf("%d %d", &n, &k);
    for (int i = 0; i < n; ++i) {
        scanf("%d", &a[i]);
    }
    a[n] = a[0];
    int cnt = 0;
    for (int i = 0; i < n; ++i) {
        cnt += (a[i] != a[i + 1]);
    }
    if (k == 1 || n == 1) {
        printf("0\n");
        return 0;
    }
    long long ans = 0;
    int freeChoices = k - 2;
    for (int i = 0; i <= cnt / 2; ++i) {
        long long cur = (cnk(cnt, i) * cnk(cnt - i, i) % MOD * powmod(freeChoices, cnt - 2 * i, MOD) % MOD * powmod(k, n - cnt, MOD) % MOD);
        ans += cur;
    }
    ans %= MOD;
    long long tot = (powmod(k, n, MOD) - ans);
    if (tot < 0) tot += MOD;
    printf("%lld\n", tot * powmod(2, MOD - 2, MOD) % MOD);
    
    return 0;
}

int main() {
    solve();
    return 0;
}
